presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseToMap.java

```diff
@@ -0,0 +1,278 @@
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.SWITCH;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.WHEN;
+import static com.facebook.presto.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.coalesce;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+import static com.facebook.presto.sql.relational.Expressions.constantNull;
+import static com.facebook.presto.sql.relational.Expressions.specialForm;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class RewriteCaseToMap
+        extends RowExpressionRewriteRuleSet
+{
+    public RewriteCaseToMap(FunctionAndTypeManager functionAndTypeManager)
+    {
+        super(new Rewriter(functionAndTypeManager));
+    }
+
+    private static class Rewriter
+            implements PlanRowExpressionRewriter
+    {
+        private final CaseToMapRewriter caseToMapRewriter;
+
+        public Rewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.caseToMapRewriter = new CaseToMapRewriter(functionAndTypeManager);
+        }
+
+        @Override
+        public RowExpression rewrite(RowExpression expression, Rule.Context context)
+        {
+            return RowExpressionTreeRewriter.rewriteWith(caseToMapRewriter, expression);
+        }
+    }
+
+    private static class CaseToMapRewriter
+            extends RowExpressionRewriter<Void>
+    {
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final FunctionResolution functionResolution;
+        private final LogicalRowExpressions logicalRowExpressions;
+
+        private CaseToMapRewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+            this.logicalRowExpressions = new LogicalRowExpressions(
+                    new RowExpressionDeterminismEvaluator(functionAndTypeManager),
+                    functionResolution,
+                    functionAndTypeManager);
+        }
+
+        private boolean addKeyValue(RowExpression key, Set<RowExpression> keySet, List<RowExpression> keys, RowExpression value, List<RowExpression> values)
+        {
+            // matching types and non-null values only allowed
+            if (!(key instanceof ConstantExpression) ||
+                    ((ConstantExpression) key).getValue() == null ||
+                    (keys.size() > 0 && !keys.get(0).getType().equals(key.getType()))) {
+                return false;
+            }
+
+            if (keySet.add(key)) {
```

Comment:
do we allow a case where we have duplicate keys? e.g.:
```
select case x when 1 then 1 when 2 then 2 when 1 then 3 end from (select 1 as x) t
```

Reply:
yes and added test